File: doc/codes/python/examples/tutorial/moving-focus-software.py
import asyncio

import numpy as np

import autd3_link_echocat as echocat
import autd3_modulation as modulation
import autd3_pattern as pattern
from autd3 import Client, ClientConfig
from autd3.commands import Pattern, SetSilencer
from autd3.geometry import Autd3, Geometry
from autd3.units import m, s
from autd3.value import Intensity

# xtask:long-running


async def main() -> None:
    geometry = Geometry([Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

    async with await Client.open(
        geometry,
        echocat.EchocatLinkOption(),
        ClientConfig(),
    ) as client:
        center = geometry.center() + np.array([0.0, 0.0, 150.0])
        wavelength = pattern.wavelength(340 * m / s)

        # ANCHOR: loop
        phases = geometry.phase_buffer()
        while True:
            for sign in (1.0, -1.0):
                target = center + np.array([sign * 20.0, 0.0, 0.0])
                pattern.focus(
                    geometry,
                    target,
                    wavelength,
                    phases,
                )
                builder = client.datagram_builder()
                builder.push(Pattern(phases, Intensity.MAX))
                for frame in builder.build():
                    await client.send_checked(frame)
                await asyncio.sleep(1.0)
        # ANCHOR_END: loop


if __name__ == "__main__":
    asyncio.run(main())

File: doc/codes/rust/examples/tutorial/moving-focus-software.rs
use std::time::Duration;

use anyhow::Result;

use autd3_rs::commands::{Modulation, Pattern, SetSilencer};
use autd3_rs::geometry::{Autd3, Geometry, offset};
use autd3_rs::units::{Hz, m, mm, s};
use autd3_rs::value::SamplingConfig;
use autd3_rs::{Client, ClientConfig};
use autd3_rs_link_nop::Nop;

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    let geometry = Geometry::new(vec![Autd3::default()]);

    let client = Client::open(&geometry, Nop, ClientConfig::default()).await?;

    let center = geometry.center() + offset(0.0 * mm, 0.0 * mm, 150.0 * mm);
    let wavelength = autd3_rs_pattern::wavelength(340.0 * m / s);

    // ANCHOR: loop
    let mut patterns = geometry.pattern_buffer();
    loop {
        for sign in [1.0_f32, -1.0] {
            let target = center + offset((sign * 20.0) * mm, 0.0 * mm, 0.0 * mm);
            autd3_rs_pattern::focus(
                &geometry,
                target,
                wavelength,
                &autd3_rs_pattern::FocusOption::default(),
                &mut patterns,
            );
            let mut builder = client.datagram_builder();
            builder.push(Pattern::new(&patterns));
            for frame in &builder.build()? {
                client.send_checked(frame).await?;
            }
            tokio::time::sleep(Duration::from_secs(1)).await;
        }
    }
    // ANCHOR_END: loop
}
